diff --git a/src/qml/qml/qqmlmetatype.cpp b/src/qml/qml/qqmlmetatype.cpp
index 89f3ced..cc3d686 100644
--- a/src/qml/qml/qqmlmetatype.cpp
+++ b/src/qml/qml/qqmlmetatype.cpp
@@ -290,23 +290,34 @@ void QQmlType::SingletonInstanceInfo::destroy(QQmlEngine *e)
     }
 }
 
+#include <QtCore/QMutex>
+#include <QtCore/QMutexLocker>
+
+namespace {
+    QMutex mutex;
+}
+
 void QQmlType::SingletonInstanceInfo::setQObjectApi(QQmlEngine *e, QObject *o)
 {
+    QMutexLocker lock(&mutex);
     qobjectApis.insert(e, o);
 }
 
 QObject *QQmlType::SingletonInstanceInfo::qobjectApi(QQmlEngine *e) const
 {
+    QMutexLocker lock(&mutex);
     return qobjectApis.value(e);
 }
 
 void QQmlType::SingletonInstanceInfo::setScriptApi(QQmlEngine *e, const QJSValue &v)
 {
+    QMutexLocker lock(&mutex);
     scriptApis.insert(e, v);
 }
 
 QJSValue QQmlType::SingletonInstanceInfo::scriptApi(QQmlEngine *e) const
 {
+    QMutexLocker lock(&mutex);
     return scriptApis.value(e);
 }
 
